### PR TITLE
Add generics to lodash.template and TemplateExecutor to allow define types to `data`

### PR DESCRIPTION
**Pull request by yihou**
Please fill in this template.

- [ x ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ x ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ x ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <https://lodash.com/docs/#template>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by yihou** [comment]
I realised whenever I use `_.template`, I can't define type for template executor's parameter, hence I added this enhancement

**Comment by yihou** [comment]
I realise whenever I use `_.template`, I can't define type for executor's parameter, hence I added this enhancement to allow defining type for TemplateExecutor's first argument.

File: types/lodash/common/string.d.ts [hunk 1]
```diff
@@ -690,21 +690,25 @@ declare module "../index" {
         template(
             string?: string,
             options?: TemplateOptions
-        ): TemplateExecutor;
+        ): TemplateExecutor<any>;
+        template<T>(
+            string?: string,
+            options?: TemplateOptions
+        ): TemplateExecutor<T>;
```
Comment (thorn0) [inline types/lodash/common/string.d.ts]: Unfortunately, it's the [common mistake](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes) number 6.

Comment (yihou) [reply types/lodash/common/string.d.ts]: Ohh.. now I see it, thanks, I think I should close this PR then.

**Pull request closed without merging**